README.md
[![CircleCI](https://circleci.com/gh/giantswarm/template.svg?style=shield&circle-token=cbabd7d13186f190fca813db4f0c732b026f5f6c)](https://circleci.com/gh/giantswarm/template)
[![GoDoc](https://godoc.org/github.com/giantswarm/badnodedetector?status.svg)](http://godoc.org/github.com/giantswarm/badnodedetector)
# badnodedetector

This is a go library to detect bad nodes on k8s cluster that should be terminated.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).



## [Unreleased]

## [3.0.0] - 2023-11-09

### Added

- Consider `DiskFull` custom node conditions created by `node-problem-detector` app to check if a node is unhealthy.

## [2.0.0] - 2022-05-16

### Changed

- Upgrade dependencies.

### Added

- Add `ResetTickCounters` to  reset counters on all nodes in cluster to zero.

## [1.0.1] - 2020-10-30

## [1.0.0] - 2020-10-26

### Added

- Add github workflows for automatic releases.
- Add bad node detector code.

[Unreleased]: https://github.com/giantswarm/badnodedetector/compare/v3.0.0...HEAD
[3.0.0]: https://github.com/giantswarm/badnodedetector/compare/v2.0.0...v3.0.0
[2.0.0]: https://github.com/giantswarm/badnodedetector/compare/v1.0.1...v2.0.0
[1.0.1]: https://github.com/giantswarm/badnodedetector/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/giantswarm/badnodedetector/releases/tag/v1.0.0


pkg/detector/detector_test.go
package detector

import (
	"context"
	"strconv"
	"testing"
	"time"

	"github.com/giantswarm/micrologger"
	"github.com/google/go-cmp/cmp"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

func Test_removeMultipleMasterNodes(t *testing.T) {
	testCases := []struct {
		name          string
		nodes         []corev1.Node
		expectedNodes []corev1.Node
	}{
		{
			name: "test 0 - 1 worker node",
			nodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
			},
			expectedNodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
			},
		},
		{
			name: "test 1 - 1 worker node, 1 master node",
			nodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
			},
			expectedNodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
			},
		},
		{
			name: "test 2 - 1 worker node, 2 master nodes",
			nodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master2",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
			},
			expectedNodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
			},
		},
		{
			name: "test 3 - 1 worker node, 3 master nodes",
			nodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master2",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master3",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
			},
			expectedNodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
			},
		},
		{
			name: "test 4 - 4 worker nodes, 3 master nodes",
			nodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker2",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master2",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker3",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master3",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker4",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
			},
			expectedNodes: []corev1.Node{
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "master1",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleMaster,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker2",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker3",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
				{
					ObjectMeta: metav1.ObjectMeta{
						Name: "worker4",
						Labels: map[string]string{
							labelNodeRole: labelNodeRoleWorker,
						},
					},
				},
			},
		},
	}

	for i, tc := range testCases {
		t.Run(strconv.Itoa(i), func(t *testing.T) {
			t.Log(tc.name)

			filteredNodes := removeMultipleMasterNodes(tc.nodes)

			if len(filteredNodes) != len(tc.expectedNodes) {
				t.Fatalf("Expected '%d' nodes but got '%d'.\n", len(tc.expectedNodes), len(filteredNodes))
			}

			if !cmp.Equal(filteredNodes, tc.expectedNodes) {
				t.Fatalf("\n\n%s\n", cmp.Diff(tc.expectedNodes, filteredNodes))

			}
		})
	}
}

func Test_maximumNodeTermination(t *testing.T) {
	testCases := []struct {
		name                         string
		nodeCount                    int
		maxNodeTerminationPercentage float64
		expectedNodeCount            int
	}{
		{
			name:                         "test 0 - basic test",
			nodeCount:                    10,
			maxNodeTerminationPercentage: 0.5,
			expectedNodeCount:            5,
		},
		{
			name:                         "test 1 - test minimal limit",
			nodeCount:                    10,
			maxNodeTerminationPercentage: 0.01,
			expectedNodeCount:            1,
		},
		{
			name:                         "test 2 - test minimal limit",
			nodeCount:                    3,
			maxNodeTerminationPercentage: 0.10,
			expectedNodeCount:            1,
		},
		{
			name:                         "test 3 - test rounding",
			nodeCount:                    10,
			maxNodeTerminationPercentage: 0.23,
			expectedNodeCount:            2,
		},
		{
			name:                         "test 4 - lot of nodes",
			nodeCount:                    1000,
			maxNodeTerminationPercentage: 0.25,
			expectedNodeCount:            250,
		},
	}

	for i, tc := range testCases {
		t.Run(strconv.Itoa(i), func(t *testing.T) {
			t.Log(tc.name)

			maxNodeCount := maximumNodeTermination(tc.nodeCount, tc.maxNodeTerminationPercentage)

			if maxNodeCount != tc.expectedNodeCount {
				t.Fatalf("Expected '%d' nodes but got '%d'.\n", tc.expectedNodeCount, maxNodeCount)
			}
		})
	}
}

func Test_isNodeUnhealthy(t *testing.T) {
	const diskFullCondition corev1.NodeConditionType = "DiskFullKubelet"

	testCases := []struct {
		name                 string
		node                 corev1.Node
		expectedNodeNotReady bool
	}{
		{
			name: "test 0 - node ready",
			node: corev1.Node{
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionTrue,
							LastHeartbeatTime: metav1.Now(),
						},
					},
				},
			},
			expectedNodeNotReady: false,
		},
		{
			name: "test 1 - node not ready",
			node: corev1.Node{
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionFalse,
							LastHeartbeatTime: metav1.Time{Time: time.Now().Add(-time.Minute * 10)},
						},
					},
				},
			},
			expectedNodeNotReady: true,
		},
		{
			name: "test 2 - not ready but only for short time",
			node: corev1.Node{
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionFalse,
							LastHeartbeatTime: metav1.Time{Time: time.Now().Add(-time.Second * 10)},
						},
					},
				},
			},
			expectedNodeNotReady: false,
		},
		{
			name: "test 3 - ready but disk full",
			node: corev1.Node{
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionTrue,
							LastHeartbeatTime: metav1.Time{Time: time.Now().Add(-time.Minute * 10)},
						},
						{
							Type:              diskFullCondition,
							Status:            corev1.ConditionTrue,
							LastHeartbeatTime: metav1.Time{Time: time.Now().Add(-time.Minute * 10)},
						},
					},
				},
			},
			expectedNodeNotReady: true,
		},
		{
			name: "test 4 - ready but disk full for a short time",
			node: corev1.Node{
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionTrue,
							LastHeartbeatTime: metav1.Time{Time: time.Now().Add(-time.Minute * 10)},
						},
						{
							Type:              diskFullCondition,
							Status:            corev1.ConditionTrue,
							LastHeartbeatTime: metav1.Time{Time: time.Now().Add(-time.Second * 10)},
						},
					},
				},
			},
			expectedNodeNotReady: false,
		},
	}

	for i, tc := range testCases {
		t.Run(strconv.Itoa(i), func(t *testing.T) {
			t.Log(tc.name)

			logger, _ := micrologger.New(micrologger.Config{})

			result := isNodeUnhealthy(context.Background(), logger, tc.node)
			if result != tc.expectedNodeNotReady {
				t.Fatalf("Expected '%t' but got '%t'.\n", tc.expectedNodeNotReady, result)
			}
		})
	}
}

func Test_nodeNotReadyTickCount(t *testing.T) {
	testCases := []struct {
		name              string
		node              corev1.Node
		expectedTickCount int
		shouldUpdate      bool
	}{
		{
			name: "test 0 - tick counter not changed - empty annotation",
			node: corev1.Node{
				ObjectMeta: metav1.ObjectMeta{},
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionTrue,
							LastHeartbeatTime: metav1.Now(),
						},
					},
				},
			},
			expectedTickCount: 0,
			shouldUpdate:      false,
		},
		{
			name: "test 1 - tick counter not changed",
			node: corev1.Node{
				ObjectMeta: metav1.ObjectMeta{
					Annotations: map[string]string{
						annotationNodeNotReadyTick: "0",
					},
				},
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionTrue,
							LastHeartbeatTime: metav1.Now(),
						},
					},
				},
			},
			expectedTickCount: 0,
			shouldUpdate:      false,
		},
		{
			name: "test 2 - tick counter increase - no annotation",
			node: corev1.Node{
				ObjectMeta: metav1.ObjectMeta{
					Annotations: map[string]string{},
				},
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionFalse,
							LastHeartbeatTime: metav1.Time{Time: time.Now().Add(-time.Minute * 10)},
						},
					},
				},
			},
			expectedTickCount: 1,
			shouldUpdate:      true,
		},
		{
			name: "test 3 - tick counter increase",
			node: corev1.Node{
				ObjectMeta: metav1.ObjectMeta{
					Annotations: map[string]string{
						annotationNodeNotReadyTick: "5",
					},
				},
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionFalse,
							LastHeartbeatTime: metav1.Time{Time: time.Now().Add(-time.Minute * 10)},
						},
					},
				},
			},
			expectedTickCount: 6,
			shouldUpdate:      true,
		},
		{
			name: "test 4 - tick counter decrease",
			node: corev1.Node{
				ObjectMeta: metav1.ObjectMeta{
					Annotations: map[string]string{
						annotationNodeNotReadyTick: "5",
					},
				},
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionTrue,
							LastHeartbeatTime: metav1.Now(),
						},
					},
				},
			},
			expectedTickCount: 4,
			shouldUpdate:      true,
		},
		{
			name: "test 5 - invalid tick counter - increase",
			node: corev1.Node{
				ObjectMeta: metav1.ObjectMeta{
					Annotations: map[string]string{
						annotationNodeNotReadyTick: "asdefg",
					},
				},
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionFalse,
							LastHeartbeatTime: metav1.Time{Time: time.Now().Add(-time.Minute * 10)},
						},
					},
				},
			},
			expectedTickCount: 1,
			shouldUpdate:      true,
		},
		{
			name: "test 6 - invalid tick counter - reset to zero",
			node: corev1.Node{
				ObjectMeta: metav1.ObjectMeta{
					Annotations: map[string]string{
						annotationNodeNotReadyTick: "asdefg",
					},
				},
				Status: corev1.NodeStatus{
					Conditions: []corev1.NodeCondition{
						{
							Type:              corev1.NodeReady,
							Status:            corev1.ConditionTrue,
							LastHeartbeatTime: metav1.Now(),
						},
					},
				},
			},
			expectedTickCount: 0,
			shouldUpdate:      true,
		},
	}

	for i, tc := range testCases {
		t.Run(strconv.Itoa(i), func(t *testing.T) {
			t.Log(tc.name)

			logger, _ := micrologger.New(micrologger.Config{})

			tickCounter, updated := nodeNotReadyTickCount(context.Background(), logger, tc.node)
			if tickCounter != tc.expectedTickCount {
				t.Fatalf("Expected tick counter '%d' but got '%d'.\n", tc.expectedTickCount, tickCounter)
			}

			if updated != tc.shouldUpdate {
				t.Fatalf("Expected tick counter updated value to be '%t' but got '%t'.\n", tc.shouldUpdate, updated)
			}

		})
	}
}


pkg/lock/timelock.go
package lock

import (
	"context"
	"fmt"
	"time"

	"github.com/giantswarm/microerror"
	"github.com/giantswarm/micrologger"
	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/types"
	"sigs.k8s.io/controller-runtime/pkg/client"
)

const (
	timeLockName = "timelock.giantswarm.io/until"
)

type Config struct {
	Logger    micrologger.Logger
	K8sClient client.Client

	Name string
	TTL  time.Duration
}

type TimeLock struct {
	logger    micrologger.Logger
	k8sClient client.Client

	name string
	ttl  time.Duration
}

// NewTimeLock implements a distributed time lock mechanism mainly used for bad node detection pause period.
// You can inspect the lock annotations on the default namespace in the TC k8s api.
// The lock is unique to each component that request the lock.
//     $ kubectl get ns default --watch | jq '.metadata.annotations'
//     "aws-operator@6.7.0.timelock.giantswarm.io/until": "Mon Jan 2 15:04:05 MST 2006"
//
func New(config Config) (*TimeLock, error) {
	if config.Logger == nil {
		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
	}
	if config.K8sClient == nil {
		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
	}
	if config.Name == "" {
		return nil, microerror.Maskf(invalidConfigError, "%T.Name must not be empty", config)

	}
	if config.TTL == 0 {
		return nil, microerror.Maskf(invalidConfigError, "%T.TTL must not be zero", config)
	}

	d := &TimeLock{
		logger:    config.Logger,
		k8sClient: config.K8sClient,

		name: config.Name,
		ttl:  config.TTL,
	}

	return d, nil
}

func (t *TimeLock) Lock(ctx context.Context) error {
	locked, err := t.isLocked(ctx)
	if err != nil {
		return err
	}

	if locked {
		// fail since lock is already acquired
		return microerror.Maskf(alreadyExistsError, fmt.Sprintf("time lock for the component %s already exists", t.name))
	}

	err = t.createLock(ctx)
	if err != nil {
		return microerror.Mask(err)
	}

	return nil
}

func (t *TimeLock) isLocked(ctx context.Context) (bool, error) {
	var err error
	isLocked := false

	var ns corev1.Namespace

	err = t.k8sClient.Get(ctx, types.NamespacedName{Name: corev1.NamespaceDefault}, &ns)
	if err != nil {
		return false, microerror.Mask(err)
	}

	timeStamp, ok := ns.Annotations[lockName(t.name)]
	if ok {
		ts, err := time.Parse(time.UnixDate, timeStamp)
		if err != nil {
			return false, microerror.Mask(err)
		}
		// check if the lock is expired
		if time.Now().Before(ts) {
			isLocked = true
		}
	}
	return isLocked, nil
}

func (t *TimeLock) createLock(ctx context.Context) error {
	var ns corev1.Namespace

	err := t.k8sClient.Get(ctx, types.NamespacedName{Name: corev1.NamespaceDefault}, &ns)
	if err != nil {
		return microerror.Mask(err)
	}
	if ns.Annotations == nil {
		ns.Annotations = map[string]string{}
	}
	// add lock timestamp
	ns.Annotations[lockName(t.name)] = time.Now().Add(t.ttl).Format(time.UnixDate)

	err = t.k8sClient.Update(ctx, &ns)
	if err != nil {
		return microerror.Mask(err)
	}

	return nil
}

func (t *TimeLock) ClearLock(ctx context.Context, component string) error {
	var ns corev1.Namespace

	err := t.k8sClient.Get(ctx, types.NamespacedName{Name: corev1.NamespaceDefault}, &ns)
	if err != nil {
		return microerror.Mask(err)
	}

	updated := false

	if _, ok := ns.Annotations[lockName(component)]; ok {
		// delete lock from annotations
		delete(ns.Annotations, component)
		updated = true
	}

	if updated {
		err = t.k8sClient.Update(ctx, &ns)
		if err != nil {
			return microerror.Mask(err)
		}
	}
	return nil
}

func lockName(component string) string {
	return fmt.Sprintf("%s.%s", component, timeLockName)
}


pkg/detector/detector.go
package detector

import (
	"context"
	"fmt"
	"math"
	"strconv"
	"time"

	"github.com/giantswarm/microerror"
	"github.com/giantswarm/micrologger"
	corev1 "k8s.io/api/core/v1"
	"sigs.k8s.io/controller-runtime/pkg/client"
)

const (
	defaultMaxNodeTerminationPercentage = 0.10
	defaultNotReadyTickThreshold        = 6
	defaultPauseBetweenTermination      = time.Minute * 10

	nodeNotReadyDuration = time.Second * 30

	annotationNodeNotReadyTick = "giantswarm.io/node-not-ready-tick"
	labelNodeRole              = "role"
	labelNodeRoleMaster        = "master"
	labelNodeRoleWorker        = "worker"
)

var trueConditions = []string{
	string(corev1.NodeReady),
}
var falseConditions = []string{
	// Custom conditionx generated by https://github.com/giantswarm/node-problem-detector-app
	"DiskFullKubelet",
	"DiskFullContainerd",
	"DiskFullVarLog",
}

type Config struct {
	Logger    micrologger.Logger
	K8sClient client.Client

	// MaxNodeTerminationPercentage defines a maximum percentage of nodes that will be returned as 'marked for termination'
	// ie: if the value is 0.5 and cluster have 10 nodes, than `DetectBadNodes`can only return maximum of 5 nodes
	// marked for termination at single run
	MaxNodeTerminationPercentage float64
	// NotReadyTickThreshold defines a how many times the node must bee seen as NotReady in order to return it as 'marked for termination'
	NotReadyTickThreshold int
	// PauseBetweenTermination defines a pause between 2 intervals where node termination can occur.
	// This is a safeguard to prevent nodes being terminated over and over or to not terminate too much at once.
	// ie: if the value is 5m it means once it returned nodes for termination it wont return another nodes for another 5 min.
	PauseBetweenTermination time.Duration
}

type Detector struct {
	logger    micrologger.Logger
	k8sClient client.Client

	maxNodeTerminationPercentage float64
	notReadyTickThreshold        int
	pauseBetweenTermination      time.Duration
}

func NewDetector(config Config) (*Detector, error) {
	if config.Logger == nil {
		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
	}
	if config.K8sClient == nil {
		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
	}

	if config.MaxNodeTerminationPercentage == 0 {
		config.MaxNodeTerminationPercentage = defaultMaxNodeTerminationPercentage
	}
	if config.NotReadyTickThreshold == 0 {
		config.NotReadyTickThreshold = defaultNotReadyTickThreshold
	}
	if config.PauseBetweenTermination == 0 {
		config.PauseBetweenTermination = defaultPauseBetweenTermination
	}

	d := &Detector{
		logger:    config.Logger,
		k8sClient: config.K8sClient,

		maxNodeTerminationPercentage: config.MaxNodeTerminationPercentage,
		notReadyTickThreshold:        config.NotReadyTickThreshold,
		pauseBetweenTermination:      config.PauseBetweenTermination,
	}

	return d, nil
}

// DetectBadNodes will return list of nodes that should be terminated which in documentation terminology is used as 'marked for termination'.
func (d *Detector) DetectBadNodes(ctx context.Context) ([]corev1.Node, error) {
	var nodeList corev1.NodeList
	{
		err := d.k8sClient.List(ctx, &nodeList)
		if err != nil {
			return nil, microerror.Mask(err)
		}
	}

	// badNodes list will contain all nodes that reached tick threshold and are 'marked for termination'
	var badNodes []corev1.Node
	for i, n := range nodeList.Items {
		notReadyTickCount, updated := nodeNotReadyTickCount(ctx, d.logger, n)

		if notReadyTickCount >= d.notReadyTickThreshold {
			badNodes = append(badNodes, n)
		}

		// if the tick counter changed, we need to update the value in the k8s api
		if updated {
			// update the tick count on the node
			n.Annotations[annotationNodeNotReadyTick] = fmt.Sprintf("%d", notReadyTickCount)
			err := d.k8sClient.Update(ctx, &nodeList.Items[i])
			if err != nil {
				return nil, microerror.Mask(err)
			}
			d.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updated not ready tick count to %d/%d for node %s", notReadyTickCount, d.notReadyTickThreshold, n.Name))
		}
	}

	// remove additional master nodes to avoid multiple master node termination at the same time
	badNodes = removeMultipleMasterNodes(badNodes)
	d.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found %d nodes marked for termination", len(badNodes)))

	// check for node termination limit, to prevent termination of all nodes at once
	maxNodeTermination := maximumNodeTermination(len(nodeList.Items), d.maxNodeTerminationPercentage)
	if len(badNodes) > maxNodeTermination {
		badNodes = badNodes[:maxNodeTermination]
		d.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("limited node termination to %d nodes", maxNodeTermination))
	}

	return badNodes, nil
}

// ResetTickCounters will reset tick counters to zero on all k8s nodes in a cluster
func (d *Detector) ResetTickCounters(ctx context.Context) error {
	var nodeList corev1.NodeList
	{
		err := d.k8sClient.List(ctx, &nodeList)
		if err != nil {
			return microerror.Mask(err)
		}
	}

	for i, node := range nodeList.Items {
		if _, ok := node.GetAnnotations()[annotationNodeNotReadyTick]; ok {
			node.Annotations[annotationNodeNotReadyTick] = "0"

			err := d.k8sClient.Update(ctx, &nodeList.Items[i])
			if err != nil {
				return microerror.Mask(err)
			}
		}
	}

	return nil
}

// isNodeUnhealthy returns true of the node is not ready for certain period of time
// this is used to detect bad nodes
func isNodeUnhealthy(ctx context.Context, logger micrologger.Logger, n corev1.Node) bool {
	// trueConditions have to be true, otherwise node has to be considered unhealthy.
	for _, trueCondition := range trueConditions {
		for _, c := range n.Status.Conditions {
			if string(c.Type) == trueCondition && c.Status != corev1.ConditionTrue {
				// We want condition to be true, but it's not.
				if time.Since(c.LastHeartbeatTime.Time) >= nodeNotReadyDuration {
					logger.Debugf(ctx, "node %s is unhealthy because we expected condition %s to be true, but was false", n.Name, c.Type)
					return true
				}
			}
		}
	}

	// falseConditions have to be false, otherwise node has to be considered unhealthy.
	for _, falseCondition := range falseConditions {
		for _, c := range n.Status.Conditions {
			if string(c.Type) == falseCondition && c.Status == corev1.ConditionTrue {
				// we want condition to be false, but it's not.
				if time.Since(c.LastHeartbeatTime.Time) >= nodeNotReadyDuration {
					logger.Debugf(ctx, "node %s is unhealthy because we expected condition %s to be false, but was true", n.Name, c.Type)
					return true
				}
			}
		}
	}
	return false
}

// updateNodeNotReadyTickAnnotations will update annotations on the node
// depending if the node is Ready or not
// the annotation is used to track how many times node was seen as not ready
// and in case it will reach a threshold, the node will be marked for termination.
// Each run of this function can increase or decrease the tick count by 1.
// function return a tick counter (int) and a bool indicating if the value changed
func nodeNotReadyTickCount(ctx context.Context, logger micrologger.Logger, n corev1.Node) (int, bool) {
	var err error
	updated := false

	// fetch current notReady tick count from node
	// if there is no annotation yet, the value will be 0
	notReadyTickCount := 0
	{
		tick, ok := n.Annotations[annotationNodeNotReadyTick]
		if ok {
			notReadyTickCount, err = strconv.Atoi(tick)
			// in case the annotation is a garbage lets reset to 0 and update it
			if err != nil {
				notReadyTickCount = 0
				updated = true
			}
		}
	}

	// increase or decrease the tick count depending on the node status
	if isNodeUnhealthy(ctx, logger, n) {
		notReadyTickCount++
		updated = true
	} else if notReadyTickCount > 0 {
		notReadyTickCount--
		updated = true
	}

	return notReadyTickCount, updated
}

// maximumNodeTermination calculates the maximum number of nodes that can be terminated on single run
// the number is calculated with help of maxNodeTerminationPercentage
// which determines how much percentage of nodes can be terminated
// the minimum is 1 node termination per run
func maximumNodeTermination(nodeCount int, maxNodeTerminationPercentage float64) int {
	limit := math.Round(float64(nodeCount) * maxNodeTerminationPercentage)

	if limit < 1 {
		limit = 1
	}
	return int(limit)
}

// removeMultipleMasterNodes removes multiple master nodes from the list to avoid more than 1 master node termination at same time
// worker nodes in the list are unaffected
func removeMultipleMasterNodes(nodeList []corev1.Node) []corev1.Node {
	foundMasterNode := false
	// filteredNodes list will contain maximum 1 master node and unlimited number of worker nodes at the end of the function
	var filteredNodes []corev1.Node

	for _, n := range nodeList {
		if n.Labels[labelNodeRole] == labelNodeRoleMaster {
			// append only the first master that is found in the list
			// any following master is not appended to the final list
			if !foundMasterNode {
				filteredNodes = append(filteredNodes, n)
				foundMasterNode = true
			} else {
				// removing additional master nodes from the list
				continue
			}
		} else {
			// append all non-master nodes
			filteredNodes = append(filteredNodes, n)
		}
	}
	return filteredNodes
}
